README.md
AWS Autoscale Combined Metrics
==============================

AWS currently (July 2015) does not evalute other policies during a cooldown peroid

meanning if CPU policy was triggered and scaling acitivty happened the other policies for Network / Memory etc.. would not be evaluated until the cooldown period for the first policy passed

this gem will push a custom metric to cloudwatch:

it will combine the metrics you want to check and only if all of them are out of the treshold you provided 

scale in if *all* metrics are in range send a 1 (ok to scale in)
scale out if *any* metrics are in range send a 1 (ok to scale out)

How it works
------------

There's a yml config file in which you specify either a cloudformation stack name and resource (i.e AppServerGroup)
in this config you specify the metrics tresholds and comparison operator to evalute (kind of alrams in CloudWatch)

you can specify sectaions for ScaleOut and ScaleIn and their related metrics in the config file

````yml
ScaleOut:
- metric_name: CPUUtilization
  namespace: AWS/EC2
  statistics:
  - Maximum
  unit: Percent
  threshold: 20
  comparison_operator: <=
  aggregate_as_group: true
````

you can specify aggregate_as_group: true in the config and it will aggregate based on the statistics you passed across ec2 instances

Note:
AWS is already agregating the instances metrics on the AutoScaleGroupName dimension for you (so if you are pushing a custom metrics you shold be pushing it to the autoscale group diemension)


if the autoscale group needs to be extracted from the stack name you specify (useful for discovery uisng cloudformation)
````yml
cloudformation:
  enabled: true
  stack_name: STACK_NAME_X
  logical_resource_ids: 
  - CloudFormation_RESOURCE_ID_1 
  - CloudFormation_RESOURCE_ID_2 
````
you can pass multiple autoscale groups given you would like to use the same metrics and tresholds for their combined metrics
 
The app will combined the results of all metrics into a true / flase array
for ScaleOut events it will check if any element in the array is true and will publish a custom metric (i.e ScaleOut_CPUUtilization_NetworkIn ) under the combined_metrics custom name space in CloudWatch on the AutoScale Group dimension you have specified in the config

you can then set a single alarm and a single policy to perform your scale activity if itsthe value is 1 (O.K to scale in/out)


you run the file through the as-combined-metrics app (a backup of the config will be created)

````bash
as-combined-metrics -f path_to/combinedMetrics.yml
````

options:
--------
````bash
  r, [--region=REGION]                 # AWS Region
                                       # Default: us-east-1
  l, [--log-level=LOG_LEVEL]           # Log level
                                       # Default: INFO
  f, [--config-file=CONFIG_FILE]       # Metrics config file location
      [--scalein-only=SCALEIN_ONLY]    # gather combined metrics for scale in only
      [--scaleout-only=SCALEOUT_ONLY]  # gather combined metrics for scale out only
  p, [--period=N]                      # Metric datapoint last x minutes
                                       # Default: 300
  t, [--timeout=N]                     # Timeout (seconds) for fetching autoscale group name
                                       # Default: 120
  o, [--once], [--no-once]             # no loop - run once
  d, [--dryrun], [--no-dryrun]         # do not submit metric to CloudWatch
  i, [--interval=N]                    # interval to check metrics
                                       # Default: 30
````
 
you may pass ---scalein-only or --scaleout-only to have this publish only scalein / scale out metric

if you choose --dryrun it will NOT publish to cloudwatch

Required privileges
-------------------

* cloudwatch:GetMetricStatistics
* cloudwatch:PutMetricData
* autoscaling:DescribeAutoScalingGroups
* cloudformation:DescribeStackResources
* cloudformation:DescribeStacks

policy sample
````json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1436799375000",
            "Effect": "Allow",
            "Action": [
                "cloudwatch:GetMetricStatistics",
                "cloudwatch:PutMetricData"
            ],
            "Resource": [
                "*"
            ]
        },
        {
            "Sid": "Stmt1436799403000",
            "Effect": "Allow",
            "Action": [
                "autoscaling:DescribeAutoScalingGroups"
            ],
            "Resource": [
                "*"
            ]
        },
        {
            "Sid": "Stmt1436799495000",
            "Effect": "Allow",
            "Action": [
                "cloudformation:DescribeStackResources",
                "cloudformation:DescribeStacks"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
````

Credentials are loaded automatically from the following locations (AWS-SDK handled):

* ENV['AWS_ACCESS_KEY_ID'] and ENV['AWS_SECRET_ACCESS_KEY']
* Aws.config[:credentials]
* Shared credentials file, ~/.aws/credentials
* EC2 Instance profile


Contributing
-------------
1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

License
-------
aws-cobined-metrics is released under [MIT License](http://www.opensource.org/licenses/MIT)


CHANGES.md
1.0.0
initial release
1.1.0
Be able to to have spot fleets monitored
1.1.1
require fileutils for use with ruby 2.6.8


lib/as-combined-metrics/cli.rb
# encoding: utf-8
$:.unshift File.expand_path(File.join(File.dirname(__FILE__), %w[.. lib]))

require 'thor'
require 'aws-sdk'
require 'yaml'
require 'pp'
require 'time'

class AsCombinedMetrics::Cli < Thor
  autoload :Logging,        'as-combined-metrics/logger'
  autoload :Config,         'as-combined-metrics/config'
  autoload :Poller,         'as-combined-metrics/poller'
  autoload :Aws,            'as-combined-metrics/aws'
  autoload :Utils,          'as-combined-metrics/utils'
  autoload :Stats,          'as-combined-metrics/stats'
  autoload :CloudWatch,     'as-combined-metrics/cloudwatch'
  autoload :CloudFormation, 'as-combined-metrics/cloudformation'

  default_task  :start

  class_option  :region,              :desc => 'AWS Region',                                          :default => 'us-east-1',      :aliases => 'r', :type => :string
  class_option  :log_level,           :desc => 'Log level',                                           :default => 'INFO',           :aliases => 'l', :type => :string
  class_option  :config_file,         :desc => 'Metrics config file location',                                                      :aliases => 'f', :type => :string
  class_option  :scalein_only,        :desc => 'gather combined metrics for scale in only',           :default => false
  class_option  :scaleout_only,       :desc => 'gather combined metrics for scale out only',          :default => false
  class_option  :period,              :desc => 'Metric datapoint last x minutes',                     :default => 300,              :aliases => 'p', :type => :numeric
  class_option  :timeout,             :desc => 'Timeout (seconds) for fetching autoscale group name', :default => 300,              :aliases => 't', :type => :numeric
  class_option  :once,                :desc => "no loop - run once",                                  :default => false,            :aliases => 'o', :type => :boolean
  class_option  :dryrun,              :desc => "do not submit metric to CloudWatch",                  :default => false,            :aliases => 'd', :type => :boolean
  class_option  :interval,            :desc => 'interval to check metrics',                           :default => 30,               :aliases => 'i', :type => :numeric
  desc "start", "combine metrics"
  def start
    %w(INT TERM USR1 USR2 TTIN).each do |sig|
      trap sig do
        puts "Got Singnal #{sig} Exiting..."
        exit 0
      end
    end

    raise Thor::RequiredArgumentMissingError, 'missing config file location [-f / --config-file]' if options[:config_file].nil?

    extend Logging
    extend Utils
    extend Config
    extend Stats
    extend Aws
    extend CloudFormation
    extend CloudWatch
    extend Poller

    logger.info { set_color "Dry Run - will not publish metrics to CloudWatch", :yellow } if options[:dryrun]
    logger.info "Starting Combined Metrics on #{options[:region]} region"
    @region = options[:region]
    init_aws_sdk
    load_config
    poll(options[:interval])
    <<-INFO
        1) in this loop (poll) we overwrite a hash of combined metrics as follow:

           for every metric in the config file use fetch_metric(metric) to get it's current reading

           if metric has a key of aggregate_as_group,
              get all instances of the AutoScale group and for each instance fetch metric
              then push the result to an array, from that array get the average result and push it to the combined_metrics hash

           if one of the metrics fails to get datapoints (result => empty datapoints set) then set the result to -1
           the check_combined_metrics will see that the value is -1 and automatically false the result


        Combined metrics:
        { "CPUUtilization" => {"measure"=>7.08, "threshold"=>30, "comparison_operator"=>">="},
          "NetworkIn" => {"measure"=>43765341.0, "threshold"=>943718400, "comparison_operator"=>">="},
          "memory_usage.heapUsed.average"=>{"measure"=> 67844551, "threshold"=>600000, "comparison_operator"=>">="}
        }

        2) we then call check_combined_metrics to check if the measure is under or above the threshold and the result (true/false) to array

        3) if all elements of that array are true we set the value of self.combined_metric_value to 1 (Scale OK) else we set it to 0 (Do not Scale)

        4) last step -> publish that value to CloudWatch
    INFO
  end

  map ["-v", "--version"] => :version
  desc "version", "version"
  def version
    say AsCombinedMetrics::ABOUT, color = :green
  end
end